library/ssl_tls.c

```diff
@@ -3635,6 +3648,27 @@ void mbedtls_ssl_session_free( mbedtls_ssl_session *session )
     mbedtls_platform_zeroize( session, sizeof( mbedtls_ssl_session ) );
 }
 
+void mbedtls_ssl_session_free( mbedtls_ssl_session *session )
```

Comment:
I would not change this function but maybe adding a check that the `next` field is equal to `NULL` when this function is called. The idea would be that we restrict the use of the `next` field to the chain of sessions whose head is `ssl->session_negotiate`. Thus it would be only when we free `ssl->session_negotiate` that we need to go through the list, not for any `mbedtls_ssl_session` object.

Reply:
If limit it to free `session_negotiate`, we have to add this point into the definition of `mbedtls_ssl_session_free` . And `mbedtls_ssl_session_free` must have a return value to let caller known what happens. 

And we have to create a function to free `session_negotiate` , it is not easy for maintenance.

Reply:
Back on this. Eventually I think that we should not change at all the function `mbedtls_ssl_session_free`. I understand that we cannot really add a check that the `next` field is equal to `NULL` as the function returns `void`. But that should not be an issue as if we free a session while it is still linked to others we will loose the reference to the end of the chain and the CI tests with memory sanitizer should catch that.

For our internal use of chain of `mbedtls_ssl_session` structures, we should have an internal function with a name that could be `mbedtls_ssl_session_free_chain` that should be called on the head of the chains of `mbedtls_ssl_session` structures: for the time being only on `session_negotiate`.